### ISSUE TITLE
Compiler passes wrong schema for a TKind

### ISSUE THREAD

**Issue by kenttregenza** [sinclairzx81/typebox#521]
In latest version 0.30.2 you have this for the new TKind function

https://github.com/sinclairzx81/typebox/blob/3a543656d6862ef4bc9eead582f1783d41b8bfc6/src/compiler/compiler.ts#L377-L379

Which calls..

https://github.com/sinclairzx81/typebox/blob/3a543656d6862ef4bc9eead582f1783d41b8bfc6/src/compiler/compiler.ts#L544C1-L548

However this means  the schema is picked up by the closure in the typeRegistryFunction is not the schema of TKind but likely a parent like a Type.Object etc

So since you are no longer using state.customs to hold the schema of the userdefined/TKind schema can I suggest to change the params of typeRegistryFunction to include the TKind schema

```ts
function* TKind(schema: Types.TSchema, references: Types.TSchema[], value: string): IterableIterator<string> {
    yield `kind('${schema[Types.Kind]}', ${value})`
}
```
and
```ts
function typeRegistryFunction(kind: string, kindSchema:TSchema, value: unknown) {
      if (!Types.TypeRegistry.Has(kind)) return false
      const checkFunc = Types.TypeRegistry.Get(kind)!
      return checkFunc(kindSchema, value)
} 
```

**Comment by sinclairzx81** [comment]
@kenttregenza Hi, Thanks for reporting.

This looks like an oversight / bug. There was some refactoring work carried out on this revision to try clean up some of the state tracking in the compiler, That `custom` variable was used to pass through the actual `TKind` (where there could be many instances of that kind), but the change was not picked up by the test suite (so I'll need to get some tests up around this)

Will resolve asap.
Cheers
S

**Comment by sinclairzx81** [comment]
@kenttregenza Heya,

The fix for this should be on `0.30.3`. Had to re-instate the `custom` variable (now called `instances`) and written some tests to specifically assert the schematics passed to the type registry function (not sure why this specific test wasn't written prior, but it's there now)

Unfortunately, couldn't go with your proposal to pass the schema through verbatim on the `typeRegistryFunction` as the code gen would need encode the schematic + TB symbols as strings (and TypeBox tries to generate non-TypeBox specific code such that users could technically implement the `kind`, `hash`, `format` functions manually without TypeBox), however have changed from using a string `key` to a numeric `ordinal` which JS might have a better time optimizing a numeric key fetch on the `instances` Map.

Thanks again for reporting! Will close off the issue for now.
Regards
S
